lib/logstash/outputs/elasticsearch/http_client.rb

```diff
@@ -103,15 +103,22 @@ def bulk(actions)
 
       return if actions.empty?
 
+      bulk_responses = []
       bulk_actions = actions.collect do |action, args, source|
         args, source = update_action_builder(args, source) if action == 'update'
 
+        unless ElasticSearch::VALID_HTTP_ACTIONS.include?(action)
+          logger.warn("Bulk request rejected because of unsupported action: ", :action => action)
+          bulk_responses << emulate_batch_error_response([[{action => args}, source].compact], 400, "unsupported action")
+          next nil
+        end
+
         if source && action != 'delete'
           next [ { action => args }, source ]
         else
           next { action => args }
         end
-      end
+      end.compact
```

Comment:
I worry that this message combined with our emulated error in the batch's combined-bulks response will cause users to assume that _Elasticsearch_ is rejecting the event, and that a batch with many such events will _flood_ the logs with identical messages. We want it to be clear that the event is being _excluded_ from being sent, and to rely on our per-event handling of errors to include event-specific context in the logs.

Speaking of which, we have an ordering problem. The return value of this method must contain its per-event responses in the [same order as it received the actions](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/1cab1ad2a41b28ac569cb77809b48ed6b66d2b39/lib/logstash/plugin_mixins/elasticsearch/common.rb#L251-L257) so that it can correctly map which events to retry.

By picking off the invalid actions here we are preventing the order from being maintained.


***

The following is what I was considering _before_ realizing that we have an ordering problem, and is _NOT_ a valid suggestion but perhaps worth including for context.

We may want to pre-partition our actions into those that are valid and those that are not valid, and emit a single message that XX events were not sent due to unsupported actions. We can rely on the downstream handling of the value we return (including the emulated error from ES) to log the individual events that are getting retried/dropped, which includes additional context about the event itself.

If I were to tackle this, I would probably break this collect block out into a private method like `build_action_tuple`, then do something like:

~~~ ruby
      bulk_actions, unsupported_actions = actions.partition do |action, _, _|
        ElasticSearch::VALID_HTTP_ACTIONS.include?(action)
      end.collect do |partition_actions|
        partition_actions.collect do |action, args, source|
          build_action_tuple(action, args, source)
        end
      end
      if unsupported_actions.any?
        logger.warn("events were excluded from batch because they had an unsupported action", event_count: unsupported_actions.size)
        bulk_responses << emulate_batch_error_response(unsupported_actions, 400, "unsupported action")
      end
~~~

Reply:
Thanks @yaauie , 
It is not really visible [the logic where maps the requested action with response](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/1cab1ad2a41b28ac569cb77809b48ed6b66d2b39/lib/logstash/plugin_mixins/elasticsearch/common.rb#L257) and I think for the long term we should avoid (maybe Hashing mechanism would be better?) to mitigate such risks.
I have updated the PR with new change and please give your feedbacks when you will have a time. Thanks.